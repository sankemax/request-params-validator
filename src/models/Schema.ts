import { ID, Method, Obj } from "./types";

export interface Schema extends Obj {
    path: string
    method: Method
}

export interface SchemaWId extends Schema {
    id: ID
}
