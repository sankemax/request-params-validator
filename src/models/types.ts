import { MissingParamError, IdError, TypeError } from "./Errors"

export interface Obj {
    [key: string]: any
}

export interface ValidationObj {
    name: string
    types: string[]
    required: boolean
}

export interface ObjectValue {
    name: string
    value: any
}

export interface Validation {
    errors: ValidationError[]
}

export interface ValidationError extends TypeError, MissingParamError, IdError {}

export interface ValidatorFn {
    name: string
    fn: Predicate
}

export interface Predicate {
    (validate: any): boolean
}

export type Method = 'POST' | 'GET' | 'PUT' | 'PATCH' // TODO: add all methods

export type ID = string
