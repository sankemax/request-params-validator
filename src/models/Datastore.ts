import { ID } from "./types";
import { Schema } from "./Schema";

export interface Datastore {
    insert(schema: Schema): Promise<ID>
    load(id: ID): Promise<Schema>
    delete(id: ID): Promise<Schema>
}
