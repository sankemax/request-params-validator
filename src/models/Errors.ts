export interface Error {
    message: string
}

export interface TypeError extends Error {
    path: string
}

export interface MissingParamError extends Error {}

export interface IdError extends Error {}
