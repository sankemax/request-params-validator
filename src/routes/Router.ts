import express, { Request, Response, NextFunction } from "express";

import mapStore from "../store/MapDatastore";
import { Schema } from "../models/Schema";
import { ID, Obj } from "../models/types";
import { validateObject } from "../validation/validate";
import { Datastore } from "../models/Datastore";

export default express.Router()
    .post('/models', modelsHandler)
    .post('/validate', validationHandler);

async function modelsHandler(req: Request, res: Response) {
    try {
        const schemas: Schema[] | Schema = req.body.data;
        const ids: ID[] = [];
        if (Array.isArray(schemas)) {
            for await (const schema of schemas) {
                ids.push(await mapStore.insert(schema));
            }
        } else {
            ids.push(await mapStore.insert(schemas));
        }

        res.json({
            success: true,
            ids,
        })

    } catch (error) {
        res.json({
            success: false,
            validationResponse: [error.message]
        })
    }
}

async function validationHandler(req: Request, res: Response) {
    try {
        const objects: Obj[] | Obj = req.body.data;
        res.json({
            success: true,
            validationResponse: await validateObjects(objects, mapStore),
        })
    } catch (error) {
        res.json({
            success: false,
            validationResponse: [error.message]
        })
    }
}

export async function validateObjects(objects: Obj[] | Obj, store: Datastore) {
    const validateObjectWithDependencies = validateObject(store);
    const response: any = [];

    if (Array.isArray(objects)) {
        for await (const object of objects) {
            const validationResponse = await validateObjectWithDependencies(object);
            response.push(validationResponse);
        }
    } else {
        const validationResponse = await validateObjectWithDependencies(objects);
        response.push(validationResponse);
    }
    return response;
}

export function internalErrorHandler(err: Error, req: Request, res: Response, next: NextFunction) {
    console.error(err.stack);
    res.json({ success: false, message: 'internal server error' });
}
