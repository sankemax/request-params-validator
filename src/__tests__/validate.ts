import models from './models.json';
import requests from './requests.json';

import { Schema } from '../models/Schema';
import mapStore from "../store/MapDatastore";
import { validateObjects } from '../routes/Router';
import { Obj } from '../models/types';

describe('validation tests', async () => {

    beforeEach(async () => {
        for await (const model of models) {
            await mapStore.insert(model as Schema);
        }
    })

    test('requests should be validated against the appropriate models', async () => {

        expect(await validateObjects(requests as Obj[], mapStore)).toMatchSnapshot()

    })

})
