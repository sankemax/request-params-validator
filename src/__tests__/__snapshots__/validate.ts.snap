// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validation tests requests should be validated against the appropriate models 1`] = `
Array [
  Object {
    "errors": Array [],
  },
  Object {
    "errors": Array [
      Object {
        "message": "Expected required field Authorization to exist.",
        "path": "/users/info-GET.headers",
      },
    ],
  },
  Object {
    "errors": Array [],
  },
  Object {
    "errors": Array [],
  },
  Object {
    "errors": Array [],
  },
  Object {
    "errors": Array [
      Object {
        "message": "Expected required field firstName to exist.",
        "path": "/users/create-POST.body",
      },
    ],
  },
  Object {
    "errors": Array [],
  },
  Object {
    "errors": Array [
      Object {
        "message": "Expected type String. Got UUID.",
        "path": "/users/create-POST.body.email",
      },
      Object {
        "message": "Expected type Email. Got UUID.",
        "path": "/users/create-POST.body.email",
      },
    ],
  },
  Object {
    "errors": Array [],
  },
  Object {
    "errors": Array [],
  },
  Object {
    "errors": Array [
      Object {
        "message": "Expected type String. Got UUID.",
        "path": "/orders/info-GET.query_params.order_id",
      },
      Object {
        "message": "Expected type UUID. Got String.",
        "path": "/orders/info-GET.query_params.order_id",
      },
    ],
  },
  Object {
    "errors": Array [],
  },
  Object {
    "errors": Array [
      Object {
        "message": "Expected type Int. Got String.",
        "path": "/orders/create-POST.body.order_type",
      },
    ],
  },
  Object {
    "errors": Array [
      Object {
        "message": "Expected required field address to exist.",
        "path": "/orders/create-POST.body",
      },
      Object {
        "message": "Expected type List. Got String.",
        "path": "/orders/create-POST.body.items",
      },
    ],
  },
  Object {
    "errors": Array [],
  },
  Object {
    "errors": Array [
      Object {
        "message": "Expected required field order_id to exist.",
        "path": "/orders/update-PATCH.body",
      },
    ],
  },
  Object {
    "errors": Array [],
  },
  Object {
    "errors": Array [
      Object {
        "message": "Expected type Int. Got String.",
        "path": "/orders/update-PATCH.body.order_type",
      },
    ],
  },
]
`;
